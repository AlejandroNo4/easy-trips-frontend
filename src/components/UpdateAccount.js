import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import fetchingPatch from '../api/fetchingPatch';
import fetchingDelete from '../api/fetchingDelete';
import FormUpdateUser from './FormUpdateUser';
import fetchingGet from '../api/fetchingGet';
import GoHomeBtn from './GoHomeBtn';

const UpdateAccount = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === false) navigate('/');
  }, []);

  const initialStateForm = {
    username: '',
    email: '',
    image: [],
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleDelete = () => {
    const url = `/users/${userState.user.id}`;
    const type = 'UI';
    fetchingDelete({
      dispatch,
      url,
      type,
    });
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const { username, email, image } = form;

    const formData = {
      user: {
        username,
        email,
        image,
      },
    };
    const url = `/users/${userState.user.id}`;
    const type = 'UI';
    fetchingPatch({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return <h1>------LOADING...------</h1>;
  }
  return (
    <div>
      <FormUpdateUser handleChange={handleChange} handleSubmit={handleSubmit} />
      <GoHomeBtn />
      {userState.user.admin === false && (
        <button type="button" onClick={handleDelete}>
          Delete account
        </button>
      )}
    </div>
  );
};

export default UpdateAccount;
