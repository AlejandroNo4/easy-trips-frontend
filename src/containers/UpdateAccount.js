import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, Link } from 'react-router-dom';
import fetchingPatch from '../api/fetchingPatch';
import fetchingDelete from '../api/fetchingDelete';
import FormUpdateUser from '../components/FormUpdateUser';
import fetchingGet from '../api/fetchingGet';

const UpdateAccount = () => {
  const dispatch = useDispatch();
  const userState = useSelector((state) => state.UIReducer);
  const navigate = useNavigate();

  useEffect(() => {
    const url = 'logged_in';
    const type = 'UI';
    fetchingGet({
      dispatch,
      url,
      type,
    });
    if (userState.user.logged_in === false) navigate('/');
  }, []);

  const initialStateForm = {
    username: '',
    email: '',
    image: {},
  };
  const [form, updateInput] = useState(initialStateForm);

  const handleChange = (e) => {
    if (e.target.files) updateInput({ ...form, [e.target.name]: e.target.files[0] });
    else updateInput({ ...form, [e.target.name]: e.target.value });
  };

  const handleDelete = () => {
    const url = `/users/${userState.user.id}`;
    const type = 'UI';
    fetchingDelete({
      dispatch,
      url,
      type,
    });
    navigate('/');
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const { username, email, image } = form;

    const formData = {
      user: {
        username,
        email,
        image,
      },
    };
    const url = `/users/${userState.user.id}`;
    const type = 'UI';
    fetchingPatch({
      dispatch,
      url,
      formData,
      navigate,
      type,
    });
  };

  if (userState.loading === true) {
    return (
      <div className="d-flex flex-column justify-center align-center w-100">
        <h1 className="session-title">Loading...</h1>
      </div>
    );
  }

  let errors;
  if (userState.loading === false) {
    const keys = Object.keys(userState.errors);
    errors = keys.map((key) => `${key}: ${userState.errors[key]}`).join('\n');
  }

  return (
    <div className="bg-no-session d-flex justify-center flex-column align-center no-session-container" data-testid="update-container">
      <p className="error-msg" data-testid="update-errs">{errors}</p>
      <h1 className="session-title">Update account</h1>
      <p className="session-description text-center">
        Please, fll out this form.
      </p>
      <FormUpdateUser
        handleChange={handleChange}
        handleSubmit={handleSubmit}
        imgSelected={form.image}
      />
      {userState.user.admin === false && (
        <button type="button" onClick={handleDelete} className="delete-account-btn">
          Delete account
        </button>
      )}
      <Link to="/" className="link-back">
        Go back
      </Link>
    </div>
  );
};

export default UpdateAccount;
