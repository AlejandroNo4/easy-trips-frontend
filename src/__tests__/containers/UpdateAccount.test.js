import React from 'react';
import { Provider } from 'react-redux';
import {
  render, screen, cleanup, fireEvent, act,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import axiosMock from 'axios';
import { BrowserRouter } from 'react-router-dom';
import store from '../../store';
import UpdateAccount from '../../containers/UpdateAccount';

afterEach(cleanup);

const renderWithRedux = (component) => ({
  ...render(
    <Provider store={store}>
      <BrowserRouter>{component}</BrowserRouter>
    </Provider>,
  ),
});

describe('UpdateAccount Component', () => {
  beforeEach(() => {
    axiosMock.get.mockResolvedValueOnce({ data: { logged_in: true } });
    renderWithRedux(<UpdateAccount />);
  });

  it('renders with Redux', async () => {
    const update = await screen.findByTestId('update-container');
    expect(update).toBeVisible();
  });

  it('matches the snapshot', async () => {
    const update = await screen.findByTestId('update-container');
    expect(update).toMatchSnapshot();
  });

  it('Renders username and tranfers the value while typing', async () => {
    const username = await screen.findByPlaceholderText('Username');
    expect(username).toBeVisible();
    userEvent.type(username, 'Mario');
    expect(username).toHaveValue('Mario');
    expect(username).not.toHaveValue('');
  });

  it('Renders email and tranfers the value while typing', async () => {
    const email = await screen.findByPlaceholderText('Email Adress');
    expect(email).toBeVisible();
    userEvent.type(email, 'anemail@mail.com');
    expect(email).toHaveValue('anemail@mail.com');
    expect(email).not.toHaveValue('');
  });

  it('Renders File Form input', async () => {
    const fileInput = await screen.findByTestId('image-update');
    expect(fileInput).toBeVisible();
  });

  it('Cannot render and cannot render data wothout input typed', async () => {
    axiosMock.patch.mockResolvedValueOnce({ data: {} });
    const username = await screen.findByPlaceholderText('Username');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('btn-update');
    expect(username).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    await act(async () => {
      fireEvent.click(button);
    });
    const message = await screen.findByTestId('update-errs');
    expect(message).toBeVisible();
  });

  it('Makes a call to fetching action when filling the fields', async () => {
    axiosMock.patch.mockResolvedValueOnce({ data: { message: 'succesfully updated' } });
    const username = await screen.findByPlaceholderText('Username');
    const email = await screen.findByPlaceholderText('Email Adress');
    const button = await screen.findByTestId('btn-update');
    expect(username).toBeVisible();
    expect(email).toBeVisible();
    expect(button).toBeVisible();
    userEvent.type(username, 'Mario');
    userEvent.type(email, 'anemail@mail.com');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(axiosMock.patch).toHaveBeenCalledTimes(1);
    expect(axiosMock.patch).not.toHaveBeenCalledTimes(0);
  });
});
