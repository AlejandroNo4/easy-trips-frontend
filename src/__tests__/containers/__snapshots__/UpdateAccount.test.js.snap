// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateAccount Component matches the snapshot 1`] = `
<h1
  class="session-title"
  data-testid="update-title"
>
  Update account
</h1>
`;
